Experimental support for dark mode for wxMSW

vadz opened the pull request:
This is not complete yet, but it's already doing something (note that you need to enable it explicitly to test, the simplest way to do it is by setting `WX_MSW_DARK_MODE=2` in the environment).

The most glaring missing piece (at least among those that I've seen, maybe there is something even worse) is support for dark scrollbars. Support for this in some other projects is achieved by hooking `OpenNcThemeData()` function and forcing it to use `Explorer::ScrollBar` instead of `ScrollBar`, but doing this requires mucking with IAT and doesn't seem very robust, so I'm not really sure what to do about it yet. We could add an option to opt-in into doing this, but this looks like a cop out rather than a real solution.

Anyhow, testing this is already welcome, even if, to return to the beginning, many things don't work yet.

Things known not to work and already in my pipeline:

- [x] Radio box buttons are unreadable (I'd like to rewrite `wxRadioBox` to use `wxRadioButton` to fix this, no idea why hadn't it been done like this 30 years ago already...).
- [x] Toolbar problems.
- [x] `wxFontPickerCtrl` initial colour.
- [x] Non-top notebook tabs.
- [ ] Light scrollbars: see above.
- [ ] wxSTC.
- [ ] wxAUI.

vadz committed:
Start adding dark mode support to wxMSW

Add experimental wxApp::MSWEnableDarkMode().

For now this is very incomplete and just uses the basic API for setting
the preferred application mode, which only works for the standard
dialogs and popup menus, but it's a start.

Support of dark mode is also limited to Windows 10 20H1 and later, but
this should, hopefully, be not a problem in practice as few people
should be using pre-2020 Windows 10 versions by now.

vadz committed:
Use dark title bar for top level windows in dark mode

Call DwmSetWindowAttribute() with the now documented (although only for
Windows 11, but clearly working for late enough Windows 10 versions)
DWMWA_USE_IMMERSIVE_DARK_MODE argument to paint title bar appropriately.

vadz committed:
Refactor code to use ShouldUseDarkMode()

Create a function checking whether we need to use dark mode or not.

No real changes.

vadz committed:
Use dark Explorer theme for the dialog gripper

This ensures that it uses correct background in dark mode, matching the
rest of the dialog.

vadz committed:
Adapt wxToolTip colours to dark mode

Use "dark Explorer" theme for this control if appropriate.

vadz committed:
Enable using dark mode for native controls

Use undocumented AllowDarkModeForWindow() and set the window theme to
"EXPLORER" even for the windows for which we don't normally do it (i.e.
not those deriving from wxSystemThemedControl), to show (at least some)
controls correctly in dark mode.

vadz committed:
Make message box title bar dark when using dark mode

Unfortunately the message box contents itself remains light, but at
least make its title bar follow the theme.

vadz committed:
Add msw.dark-mode system option to allow turning it on easily

It doesn't make much sense to set this option programmatically, as the
application could just call MSWEnableDarkMode() directly instead, but
it's useful to have to allow enabling dark mode for any wx application
by just setting WX_MSW_DARK_MODE environment variable.

vadz committed:
Add wxMSWDarkMode::IsActive()

This is the same as the already existing wxMSWImpl::ShouldUseDarkMode()
but has a better name for the use outside of the file implementing dark
mode support.

vadz committed:
Add support for using custom theme class in dark mode

This is needed for at least wxStatusBar, although the gripper is still
not drawn correctly even when this particular theme class is used -- but
at least the main part of the status bar now uses the correct colours.

vadz committed:
Support dark mode for windows and dialogs backgrounds

Currently the colours used in the dark mode are hard-coded, which
definitely isn't great, but better than nothing and at least allows
using the correct background colour for generic windows, panels and
dialogs.

Also return dark mode-specific colours from wxSystemSettings when it's
active.

vadz committed:
Document wxMessageBox dark mode limitation

The native task dialog just doesn't seem to support dark mode and it
doesn't seem possible to change its colours manually neither.

vadz committed:
Add MSWShouldUseAutoDarkMode() and use it for wxSpinButton

This control doesn't seem to support dark mode natively, whichever theme
is used for it, so paint it in dark mode ourselves simply by inverting
its colours -- this is not great, but passable and better than leaving
it with a light background.

As this probably won't be the only control which will need this,
implement it at wxWindow level and allow the derived classes to opt-in
this "automatic dark mode support" by simply overriding the new
MSWShouldUseAutoDarkMode() function.

vadz committed:
Allow constructing wxUxThemeHandle from an HWND too

This is not used used yet, but will be soon for getting theme data of
the header control which is part of wxListCtrl but doesn't have any
corresponding wxWindow. It's also useful for using null HWND, which
must be used with the special "ItemsView" theme when using dark mode.

vadz committed:
Make wxUxThemeHandle::m_hTheme const

No real changes, just make a member variable which never changes const.

vadz committed:
Allow customizing both dark theme name and ID

We're going to need to use "ItemsView" theme for the header and list
controls, so replace the previously existing MSWGetDarkThemeClass() with
two functions, allowing to customize the name and the ID independently.

No changes yet.

vadz committed:
Enable dark mode support for wxListCtrl

Use "ItemsView" theme which draws the items of the control itself in the
same way as Explorer does when using dark mode.

Unfortunately this theme doesn't draw the scrollbars correctly, but this
will hopefully be fixed with another hack later.

vadz committed:
Apply dark mode to the tooltips of the native controls using them

Override the newly added MSWGetToolTipMessage() in the controls using
tooltips and apply dark mode to the window returned by this message too
to make sure they appear correctly in dark mode.

vadz committed:
Reuse wxMSWHeaderCtrlCustomDraw in wxListCtrl too

It doesn't seem to make much sense to have a completely identical
wxMSWListHeaderCustomDraw class used in wxListCtrl when we could simply
reuse wxMSWHeaderCtrlCustomDraw already existing elsewhere, so just do
this now.

No real changes, this is just a refactoring.

vadz committed:
Add wxUxThemeHandle::GetColour() replacing MSWGetThemeColour()

Add a helper function wrapping ::GetThemeColor() and use it in all
places where this function was used before, as it's more convenient.

Also remove wxWindow::MSWGetThemeColour() as it was only used in a
single place and doesn't seem to present any advantages compared to
using wxUxThemeHandle new method.

There should be no changes in behaviour, this is just a refactoring.

vadz committed:
Add dark mode support to wxHeaderCtrl

Extract the code setting up the header control colours for the dark mode
from src/msw/listctrl.cpp into wxMSWHeaderCtrlCustomDraw and reuse it in
wxHeaderCtrl too.

vadz committed:
Rename WindowHDC private helper class to ClientHDC

It was pretty confusing that WindowHDC actually used client HDC and not
the (full) window one.

Also add a helper AutoHDC class releasing the DC automatically in its
dtor and use it as the common base for ScreenHDC and ClientHDC.

No real changes, this is just a renaming/refactoring.

vadz committed:
Add wxMapWindowPoints() private helper function

This allows to avoid having to use casts from RECT to POINT in the rest
of the code by encapsulating this cast, and the comment explaining it,
in this function only.

No real changes.

vadz committed:
Add WindowHDC RAII helper class

After the renaming in the parent commit, this name is now free to be
used for the class creating a window HDC.

vadz committed:
Use custom draw for the menu bar in dark mode

This requires using a couple of undocumented messages/structs, which is
not ideal, but seems to be the only way to get dark menu bar and so
seems to be worth it, as white menu bar looks really out of place in
dark mode.

vadz committed:
Use the same background for windows and frames in dark mode

Redefine wxSYS_COLOUR_BTNFACE to be the same as wxSYS_COLOUR_WINDOW as
this is what windows seem to be actually using by default in the dark
mode, i.e. now wxWindow::GetBackgroundColour() returns the actual colour
being used.

vadz committed:
Replace a pair of functions with single MSWGetDarkModeSupport()

Overriding either MSWGetDarkThemeName() or MSWGetDarkThemeId() was a bit
confusing as doing the latter changed the theme name used by default too
and not flexible enough, so remove these two functions and add a single
MSWGetDarkModeSupport() one that returns both the theme name and ID and
can also be extended to do more things in the future.

vadz committed:
Fix static box and check/radio buttons text colour in dark mode

The labels of these controls were unreadable because they used the same
text colour as in normal, light mode even after enabling dark mode
support for them, so fix this by explicitly setting the default (for
dark mode, i.e. light) text colour for them.

Add a new "setForeground" field to MSWDarkModeSupport struct to allow
doing this easily and in a single place in wxControl code, so that it
could be customized, or even skipped, if these controls support for the
dark mode is improved, in the future.

vadz committed:
Document that date/time picker controls don't support dark mode

There doesn't seem to be any way to change the appearance of these
controls, DTM_SETMCCOLOR only works when not using themes.

vadz committed:
Make wxActivityIndicator visible when using dark mode

Use white colour for the spinning dots on the dark background instead of
always using black for them.

vadz commented:
It seems to be simply impossible to customize appearance of some native controls, e.g. date/time pickers don't seem to allow doing it in any way and using `MSWShouldUseAutoDarkMode()` hack for them is just too ugly -- even using the generic versions would be better.

The radio box buttons are also using black text on dark background right now, but this can be fixed by just (finally) rewriting `wxRadioBox` to use `wxRadioButton`s inside it instead of using the native buttons directly and I plan to do it.

There are certainly other problems in addition to these ones, but it could still be already interesting to test this branch with the real applications. Please let me know if anybody does it!

computergeek1507 commented:
I started testing the dark mode code on the xLights project, The AUI controls don't change to dark mode in my testing. Any ideas how to fix this. Thanks for all your hard work.

![image](https://user-images.githubusercontent.com/2293065/207915788-cb0475d7-caa9-4556-b091-fb92f85a2249.png)

vadz commented:
Thanks for testing and I should have mentioned that this part is only about the native Windows controls (and some generic ones, such as `wxActivityIndicator`). I still need to do it for wxAUI and wxSTC too. Hopefully I'll be able to do it soon, but I haven't done it yet.

eranif commented:
Hi, I tried using this branch and enabled `wxSystemOptions::SetOption("msw.dark-mode", 2);` in the dataview.cpp sample file.
No apparent change, what am I missing?

```diff
diff --git a/samples/dataview/dataview.cpp b/samples/dataview/dataview.cpp
index 67314a3251..25389e8f28 100644
--- a/samples/dataview/dataview.cpp
+++ b/samples/dataview/dataview.cpp
@@ -32,6 +32,7 @@
 #include "wx/imaglist.h"
 #include "wx/itemattr.h"
 #include "wx/notebook.h"
+#include "wx/sysopt.h"

 #ifdef wxHAS_GENERIC_DATAVIEWCTRL
     #include "wx/headerctrl.h"
@@ -380,6 +381,7 @@ wxIMPLEMENT_APP(MyApp);

 bool MyApp::OnInit()
 {
+    wxSystemOptions::SetOption("msw.dark-mode", 2);
     if ( !wxApp::OnInit() )
         return false;
```

vadz commented:
Calling this option here is too late, but then this mechanism is mostly meant to be used to force the dark mode from outside of the program, if you want to enable it programmatically, just call `MSWEnableDarkMode()` directly.

Also notice that there is a known problem with the wxDVC selection colour in dark mode that I'll fix soon.

vadz committed:
Remove unneeded headers inclusion from src/msw/notebook.cpp

No real changes, just don't include the headers which are either already
guaranteed to be included or not needed at all.

vadz committed:
Use proper theme for dark mode support in wxChoice and wxComboBox

The right theme to use is apparently "CFD".

eranif commented:
Thanks - I was able to get it to work.

**Major problem (IMO)**

- The Dark mode ON and OFF can not be done without restarting the program. My current implementation (Many of the controls I am using are custom drawn) allows for switching between dark/light modes without restart required based on the editor (wxSTC based control) selected theme

**Observations**

- `wxToolBar` (while in a wxDialog) shows black text over a dark background - which makes it unreadable. The hover background colour can not be seen as well
- Anything within a `wxNotebook` inherits its **light**  background, while the text foreground is using **light** colour as well, as a result, most of the controls looks "disabled" (see below screenshots)
- `wxFontPicker` button shows black colour over black background
- nit: The checkbox are too light while using the dark mode
- `wxSTC ` are not using dark mode scrollbars -> in my current code, I have managed to get it calling recursively (see code example below)
- `wxDialog` resize bitmap does not fit well with the dark mode colours
- Many controls border are too shiny (too white) - which produce a high contrast colours
- `wxStaticLine` colour is too shiny as well (screenshot below)

```c++
m_pfnAllowDarkModeForWindow(w->GetHandle(), use_dark);
SetWindowTheme(w->GetHandle(), use_dark ? L"DarkMode_Explorer" : L"Explorer", nullptr);
```

Below code is the current code (before your effort for Dark Mode) that I have been using until now and it also
fixes dark scrollbars:

```c++
        // bfs the windows
        std::vector<wxWindow*> Q;
        std::unordered_set<wxWindow*> V; // visited
        Q.push_back(win);
        while(!Q.empty()) {
            wxWindow* w = Q.front();
            Q.erase(Q.begin());

            if(!V.insert(w).second) {
                // already visited this window (how can this be true??)
                continue;
            }

            BOOL use_dark = current_theme_is_dark ? TRUE : FALSE;
            if(dynamic_cast<wxTextCtrl*>(w)) {
                use_dark = FALSE; // don't allow dark mode for text controls
            }

            m_pfnAllowDarkModeForWindow(w->GetHandle(), use_dark);
            SetWindowTheme(w->GetHandle(), use_dark ? L"DarkMode_Explorer" : L"Explorer", nullptr);

            if(m_pfnFlushMenuThemes) {
                m_pfnFlushMenuThemes();
            }

            InvalidateRect(w->GetHandle(), nullptr, FALSE); // HACK
            const auto& children = w->GetChildren();
            for(auto c : children) {
                Q.push_back(c);
            }
        }
```

Some screenshots to demonstrate most of the problems I mentioned (I have many dialogs in CodeLite, I can post more):

![dark-mode-1](https://user-images.githubusercontent.com/3677711/208847826-096939e7-28d4-4eda-bb36-8cbeb5af34c6.png)
![dark-mode-2](https://user-images.githubusercontent.com/3677711/208847834-b04308ea-6911-4810-981c-688ef7262b4e.png)
![dark-mode-3](https://user-images.githubusercontent.com/3677711/208847837-77df0f4e-2779-48ff-9c88-6a622476d9a4.png)

vadz commented:
Thanks a lot for testing, but we probably need to create separate issues for the specific problems as there are so many of them already (here and in [another reply](https://groups.google.com/g/wx-dev/c/ZkiLFd0zhKc/m/dD5IlvA5BgAJ) which didn't get copied here) that some of them risk getting lost anyhow.

Right now I'm still fixing `wxNotebook`, which is painful because apparently I have to draw it entirely on my own to achieve the dark look, but I'm especially interested in the scrollbar part: how exactly did you manage to make this work? Is just setting "DarkMode_Explorer" theme enough? I'm almost sure I've tried doing this and it didn't work, but perhaps I made some mistake, so I'll try again once I finish with the notebook. I'm really confused by how themes work with `AllowDarkModeForWindow()`, mostly I don't see any difference between using "DarkMode_Explorer" and "Explorer"...

vadz committed:
Start implementing our own wxNotebook drawing in dark mode

So far only tabs on the top are supported, but this will be generalized
later.

vadz commented:
@eranif Specifically concerning `wxStaticLine`, I can confirm that I don't see any difference between using "DarkMode_Explorer" and "Explorer" theme, i.e. it's still white (0xffffff) for me with the former theme. I guess we could owner-draw it to make it of the expected colour (0xe0e0e0 I'd guess?), but I still don't understand why does it work differently for you.

I'm testing under 10.0.19043.2251, what is your OS version?

eranif commented:
I will try and make some changes to the dataview demo and see what I can come up with

eranif commented:
Turns out that another call to `AllowDarkModeForWindow` was missing. I thought that you are already doing this internally so I am not sure why the extra call is needed.

Patch:
[dataview.patch](https://github.com/wxWidgets/wxWidgets/files/10284800/dataview.patch)


By adding a call to `AllowDarkModeForWindow` for every `wxDataViewCtrl` we have in the sample, gives us dark scrollbars.

Without the patch:
![normal-scrollbars](https://user-images.githubusercontent.com/3677711/209082727-4c78503e-6509-4e79-876e-35416dfc997e.png)

With the extra call:

![dark-scrollbars](https://user-images.githubusercontent.com/3677711/209082317-ac30da88-f718-45c4-9583-2a71b6413e50.png)

eranif commented:
I did some more digging and I wanted to fix the following:
- Tree buttons are not visible in dark mode
- Text selection is still behaving like it is a "light themed" control

So what I did is:
- Call `SetWindowTheme(win->GetHWND(), useDarkMode ? L"DarkMode_Explorer" : L"Explorer", nullptr);` **recursively** on all the children of the `wxDataViewCtrl`
- This fixes the problem with one caveat: `wxHeaderCtrlBase` does not play nice with `DarkMode_Explorer` - so I excluded it to get a perfect `wxDataViewCtrl`

![even-better](https://user-images.githubusercontent.com/3677711/209086228-016e1424-f425-44fe-a52e-af9d6e0452a5.png)

The patch:
[dataview.patch](https://github.com/wxWidgets/wxWidgets/files/10284918/dataview.patch)

vadz commented:
Sorry, the scrollbars for our own windows can indeed be made dark by just allowing the use of dark mode for them and setting "Explorer" theme ("DarkMode_Explorer" works too, but I still don't see any difference with it compared to "Explorer" and I think that using "Explorer" might be a tad more reliable, i.e. less likely to get broken in some future version of Windows). It's just the scrollbars of the LISTVIEW that are problematic because if we use "Explorer" for it, it uses wrong selection colour, see the comment in `wxListCtrl::MSWGetDarkModeSupport()`.

vadz committed:
Enable using dark scrollbars for all windows

Even non-native windows still need wxMSWDarkMode::AllowForWindow() to be
called in dark mode, as without it their scrollbars would use light
colours.

vadz committed:
Define wxSYS_COLOUR_BTNSHADOW and HOTLIGHT for dark mode too

HOTLIGHT is used by Explorer in dark mode to highlight its "hot" toolbar
buttons and BTNSHADOW is just a symbolic name for black, when we need a
colour even darker than WINDOW.

vadz committed:
Paint wxNotebook tabs in all directions in dark mode

Add new DrawNotebookTab() which is currently local to this file, but
it's still better to make it separate from MSWNotebookPaint() to
facilitate testing and maybe move it to wxRendererNative in the future,
if we provide an actually native implementation of it.

vadz committed:
Use theme-appropriate default colour in wxFontPickerCtrl

Don't hardcode black as the default font colour, as this can be
unreadable in dark mode.

Also update the unit test to not always expect the initial colour to be
black.

vadz committed:
Improve wxToolBar appearance in dark mode

Fix labels and highlight colour and get rid of the white line on top of
the toolbar.

Unfortunately drop-down items are still not drawn correctly as I
couldn't find any way to customize the colour used for their arrow.

vadz committed:
Allow using wxMSWDarkMode::PaintIfNecessary() with any HWND

We don't always have a wxWindow to pass to this function, so make it
more general at the cost of being slightly more inconvenient to use.

No real changes yet.

vadz committed:
Invert the spin control used by wxNotebook in dark mode

Paint it in inverted colours just as we do for standalone wxSpinButton.

This still doesn't look good, but also still looks better than nothing.

vadz commented:
I've pushed the latest round of changes. There are still many problems remaining, but I think we should merge what we have already, unless somebody really objects to it, and open separate issues/PRs for the other issues as there are definitely still going to be many of them and we'll never merge this if we wait until everything is fixed.

k-serg commented:
> I've pushed the latest round of changes. There are still many problems remaining, but I think we should merge what we have already, unless somebody really objects to it, and open separate issues/PRs for the other issues as there are definitely still going to be many of them and we'll never merge this if we wait until everything is fixed.

wxStaticBoxSizer inside wxStaticBoxSizer in dark mode not visible at all.

![Image 2022-12-23 21 19 55](https://user-images.githubusercontent.com/76852271/209376000-61ede1f4-ca2d-4f41-8e32-472183e50648.png)
![Image 2022-12-23 21 19 11](https://user-images.githubusercontent.com/76852271/209376004-c804b2c2-6087-4e1c-81e4-a8499e5c693f.png)

vadz commented:
@k-serg I can't reproduce this with the following code in the minimal sample:
```cpp
    auto p = new wxPanel(this);
    auto* s = new wxStaticBoxSizer(wxHORIZONTAL, p, "&Outer");
    auto* s2 = new wxStaticBoxSizer(wxVERTICAL, s->GetStaticBox(), "&Inner");
    s2->Add(new wxCheckBox(s2->GetStaticBox(), wxID_ANY, "&Check me"));
    s->Add(s2);
    p->SetSizer(s);
```

I've tried using the panel because I couldn't reproduce it without it neither, i.e. if you remove `p` and use `this` instead of it, everything still looks fine here.

Could you please provide a minimal patch to the minimal sample allowing to reproduce the above? TIA!

eranif commented:
I think its a good idea to push the current state into `master`. Once merged, I can perform more checks and report back

k-serg commented:
> @k-serg I can't reproduce this with the following code in the minimal sample:
> 
> ```c++
>     auto p = new wxPanel(this);
>     auto* s = new wxStaticBoxSizer(wxHORIZONTAL, p, "&Outer");
>     auto* s2 = new wxStaticBoxSizer(wxVERTICAL, s->GetStaticBox(), "&Inner");
>     s2->Add(new wxCheckBox(s2->GetStaticBox(), wxID_ANY, "&Check me"));
>     s->Add(s2);
>     p->SetSizer(s);
> ```
> 
> I've tried using the panel because I couldn't reproduce it without it neither, i.e. if you remove `p` and use `this` instead of it, everything still looks fine here.
> 
> Could you please provide a minimal patch to the minimal sample allowing to reproduce the above? TIA!

Just standart widgets.exe from \samples\widgets without any changes from my part:

![Image 2022-12-24 17 36 03](https://user-images.githubusercontent.com/76852271/209438638-b2024498-c84f-44a7-92b8-dd9882308d02.png)
![Image 2022-12-24 17 35 39](https://user-images.githubusercontent.com/76852271/209438640-226f865d-a2b4-4c0e-95e0-8451b8e8a692.png)

Also I've found thats some scrollbar again not using dark mode: widgets from \samples\widgets

![Scrollbar](https://user-images.githubusercontent.com/76852271/209438723-40e1419c-ca52-4cd4-8047-3cceec238a80.png)

k-serg commented:
And from our previous mailing list conversation:

https://groups.google.com/g/wx-dev/c/ZkiLFd0zhKc/m/RpOktT6cAwAJ

>Resizable border in Dialogs or frames with wxRESIZE_BORDER looks strange 

>Hmm, I don't see anything obviously wrong here, what exactly do you think
is strange here?

Please take a look at the color of resizeble (dotted) area. I'ts not a really big problem, but anyway.
![Resize_](https://user-images.githubusercontent.com/76852271/209439643-441728f0-de78-4e99-90f3-d43f40d4f789.png)

minimal from \samples\widgets without changes:

![Image 2022-12-24 18 12 58](https://user-images.githubusercontent.com/76852271/209439662-289d4910-7c0d-42d8-af82-f967d48cf1fb.png)

Check it please, maybe this is from my part.

vadz commented:
The trouble is that we create a scrollbar control (with `SBS_SIZEGRIP` style) ourselves for the dialog, while for the frame we just use the standard status bar which does what it does in dark mode (enabled by using "ExplorerStatusBar" for it) and even though they both seem to work on their own, they are inconsistent.

I also have no idea how is the gripper supposed to look like in dark mode. Explorer itself doesn't use it and I couldn't find any other dark mode application with it.

eranif commented:
I would vote for not having a gripper at all if you can not find a suitable dark one
On other platforms, there is no gripper... (and in most Windows native frames)

vadz commented:
It's a style, so if you don't want to use it you already can -- just don't specify `wxSTB_SIZEGRIP`, but it's a bit weird not to use it if this style is given. Maybe we shouldn't use it by default...

vadz commented:
@k-serg Note that the problems in the widgets sample are specific to using `wxRadioBox` and are addressed by #23075 (which is not really dark mode-specific, which is why I created a separate PR for it, but fixes this).

k-serg commented:
> @k-serg Note that the problems in the widgets sample are specific to using `wxRadioBox` and are addressed by #23075 (which is not really dark mode-specific, which is why I created a separate PR for it, but fixes this).

I'll try to figure out what's wrong with wxStaticBoxSizer from my part and come back tomorrow with an example (if I figure it out). And thanks again for your excellent work.

vadz committed:
Fix wxChoice and wxComboBox drop down scrollbars in dark mode

Explicitly enable dark mode for them too, as otherwise they stay light.

vadz commented:
I've force pushed this branch to apply the fixups and will merge it relatively soon if there are no objections.

k-serg commented:
Another one: wxRichToolTip painted in default (light) colors.  
 It's not big problem and easily fixed by:

if (wxMSWDarkMode::IsActive())
{
  tip.SetBackgroundColour(wxSystemSettings::GetColour(wxSYS_COLOUR_WINDOW));
}
![wxRichToolTip](https://user-images.githubusercontent.com/76852271/209455431-0b31b7a6-a2c7-4994-a1bf-5bd684566925.png)

vadz commented:
Thanks, this is indeed easy to fix and I will, but there is also the native tooltip shown for the text control that uses `Edit_ShowBalloonTip()` currently and I don't see how can we change the colours used for that one at all, so I'm not sure what to do about it -- we could either keep using the native tip with the wrong colours or use our tooltip with better colours but not quite the same appearance.

vadz committed:
Add wxSystemSettings::SelectLightDark() and use it

Add a convenient helper allowing to select between the two colours
depending on whether we're using light or dark mode.

Use this in a couple of places instead of manual tests for IsDark().

No real changes.

vadz committed:
Fix wxRichToolTip background colour in dark mode

Hardcoded colours made the text unreadable in dark mode, so use
different (even though still hardcoded) defaults for it.

k-serg commented:
> @k-serg I can't reproduce this with the following code in the minimal sample:
> 
> ```c++
>     auto p = new wxPanel(this);
>     auto* s = new wxStaticBoxSizer(wxHORIZONTAL, p, "&Outer");
>     auto* s2 = new wxStaticBoxSizer(wxVERTICAL, s->GetStaticBox(), "&Inner");
>     s2->Add(new wxCheckBox(s2->GetStaticBox(), wxID_ANY, "&Check me"));
>     s->Add(s2);
>     p->SetSizer(s);
> ```
> 
> I've tried using the panel because I couldn't reproduce it without it neither, i.e. if you remove `p` and use `this` instead of it, everything still looks fine here.
> 
> Could you please provide a minimal patch to the minimal sample allowing to reproduce the above? TIA!

Well... I've found a problem,
I've modified vadz samle:

```
    auto p = new wxPanel(this);
    auto* s = new wxStaticBoxSizer(wxHORIZONTAL, p, "&Outer");
    auto* s2 = new wxStaticBoxSizer(wxVERTICAL, p, "&Inner");
    s2->Add(new wxCheckBox(s2->GetStaticBox(), wxID_ANY, "&Check me"));
    s->Add(s2);
    p->SetSizer(s);
```
s and s2 disapear, like I said. 

I realy don't understand, why s->GetStaticBox(), instead of p. 
We use wxStaticBoxSizer(int orient, wxWindow *win, const wxString& label = wxEmptyString); where parent p (should be) the same, for s and s2, isn't?

k-serg commented:
Nevermind (my mistake), I need to get some sleep, then investigate what's wrong on my part with wxStaticBox.

vadz committed:
Don't return true from wxSystemAppearance::IsDark() in light mode

Even if the system is configured to use the dark mode, we shouldn't
return true from this function unless the application itself uses dark
mode, otherwise we're going to use wrong colours not providing enough
contrast with the system-provided ones.

This basically reverts #22020 as now that we have real dark mode
support, we should be using it instead.

k-serg commented:
@vadz Assert on wxDirPickerCtrl:

![assert](https://user-images.githubusercontent.com/76852271/209479571-970d8775-a1ab-4a07-8d68-3f12950a2411.png)

src/msw/darkmode.cpp [267]:


```
        if ( FAILED(hr) )
        {
            wxLogApiError(wxString::Format("SetWindowTheme(%08x, %s, %s)",
                                           hwnd, themeName, themeId), hr);
        }
```
Checked on my own project, and samples\widgets\.
Can you confirm?

vadz commented:
Yes, thanks. There are several problems here, but they're all minor/not really related to the dark mode, I'll push fixes soon.

vadz committed:
Don't try creating the gripper before wxDialog is created

Don't do anything in wxDialog::SetWindowStyleFlag() if it's called
before the dialog is actually created, as it happened during wxDirDialog
initialization.

Gripper creation failed anyhow in this case and this also resulted in a
debug message from wxMSWDarkMode::AllowForWindow() as we passed it an
invalid HWND.

k-serg commented:
@vadz 
I have a question: in my application I'm using wxSystemSettings::GetAppearance().wxSystemAppearance::IsDark()
for checking for system-wide enabled dark mode. When user run application for the first time (and application doesn't have config) 
I check system-wide settings like this:

    bool isDarkModeDefault = wxSystemSettings::GetAppearance().wxSystemAppearance::IsDark();
    //Check application config itself, if entry exists , i.e. user defined this, use config, if not defined use isDarkModeDefault as default value for isDarkMode.
    bool isDarkMode = SaveGameProConfig::Get()->GetParameter(wxT("/Appearance"), wxT("DarkMode"), isDarkModeDefault);
    if (isDarkMode)
    {
        EnableDarkMode(isDarkMode);
    }

And depend on what IsDark() returned run application in dark or default mode.
Not only frames/dialogs/etc itself, but also icons (I parse svg's and replace colors in runtime, then make bitmaps from svg's). 
Also application performs some custom drawing stuff, like dialog headers, depend on colors, that depend on enabled dar mode or not enabled.

Like this:
![Image 2022-12-26 12 00 30](https://user-images.githubusercontent.com/76852271/209552541-9b750121-3fab-4865-b937-6bb13563ce6e.png)
![Image 2022-12-22 07 52 37](https://user-images.githubusercontent.com/76852271/209552593-e99105a7-628e-41f4-adf3-4e7799ca6eb4.png)

Until 32daebe it works like expected: wxSystemAppearance::IsDark() return correct values for system-wide dark mode. And application run as expected in light/dark mode.
But now it always return false, cause this:

"we shouldn't
return true from this function unless the application itself uses dark
mode"

The question is: why we have wxSystemAppearance::IsDark() and wxMSWDarkMode::IsActive() which both pointed us to application dark mode? Is this wxSystemAppearance shouldn't pointed us to system settings, not application?
And second: how we may check system-wide dark mode by API right now (not registry entry "AppsUseLightTheme" Software\\Microsoft\\Windows\\CurrentVersion\\Themes\\Personalize).

vadz commented:
There is some confusion about what `wxSystemAppearance::IsDark()` means. In a lot of code I've seen (I had done code search for it), including our own code in `wxGrid` and wxAUI, it was clearly used to check if this application appearance was dark, as it's used to choose the contrasting colours and it doesn't make sense to use e.g. white foreground unless the background of this application is black, irrespectively of what the rest of the system is using.

However there is also your use case that I hadn't thought about (sorry) but which does make sense too.

I see 2 possibilities here:

1. Revert my latest change and really return true from `wxSystemAppearance::IsDark()` if the system appearance is dark, even if the application appearance is light.
2. Keep this change and add some new `wxApp::MSWShouldAppsUseDarkMode()`.

(1) has the advantage of being backwards-compatible, but doesn't make much sense otherwise, both for the reasons above (most code using this function really wants to check if the application is using dark mode, not the system) and also because this is not how this function behaves under Mac (where it uses [NSApp.effectiveAppearance](https://developer.apple.com/documentation/appkit/nsapplication/2967171-effectiveappearance?language=objc) which is also app-specific) nor under all the other platforms (where it uses `IsUsingDarkBackground()` which is based on system colours for this application). Also, if we do (1), we need some way of checking whether this application is using dark mode and I have no idea what could we call it without creating a lot of confusion, i.e. how are we going to explain that some `wxSystemSettings::IsAppDark()` should be used instead of `wxSystemAppearance::IsDark()`.

So I think we need to add another function for querying the system appearance and I think that this function should be MSW-specific because it's the only platform where we can change this (by calling `MSWEnableDarkMode()`) anyhow.

What do you think?

k-serg commented:
@vadz 
> What do you think?

I see solution that not break existing codebase but logically use wxSystemAppearance as expected (from my humble point of view) by adding:

`wxSystemAppearance::AppsUseDark()` - for aplicattions (registry entry "AppsUseLightTheme")

and

`wxSystemAppearance::SystemUseDark()` - Windows itself (registry entry "SystemUsesLightTheme": taskbar, start menu, etc)

(maybe with MSW prefixes -  `MSWAppsUseDark()`, `MSWSystemUseDark()` respectively).

I think this clearly pointing to the right place without any confusion. Word "Should" not the best (from my taste): in various use case we just check system appearance, not force todo something.

wxApp (again, on my taste), not good place for it: wxApp pointing to something like application behaviour. 

Maybe I'm wrong, maybe not, just opinion.

vadz commented:
I'm fine with adding new methods to `wxSystemAppearance` rather than `wxApp`, the real question to me is whether we change the existing `IsDark()` in a backwards-incompatible way or not.

So do we all agree on:

1. `IsDark()`: is this application itself using a dark theme.
2. `AreAppsDark()`: do the applications use dark theme by default (same as `IsDark()` under non-MSW for now but could change if somebody figures out how to do it).
3. `IsSystemDark()`: does the system UI use dark theme (same as `AreAppsDark()` under non-MSW and will most likely remain this way).

?

k-serg commented:
I don't know what others will say, but it seems pretty acceptable to me.

oneeyeman1 commented:
You have been asked to test it to make sure it really fixed. 😀

vadz committed:
Add wxSystemAppearance::AreAppsDark() and IsSystemDark()

This allows to determine if dark mode is enabled (for the applications
and/or the system UI itself, respectively) system-wide which is useful
now that IsDark() returns false unless MSWEnableDarkMode() is called.

Also document the incompatible change to IsDark().

vadz committed:
Send wxEVT_SYS_COLOUR_CHANGED on dark/light theme change in wxMSW

wxOSX already does it, so do it in wxMSW too for consistency and because
it seems useful to be able to react to the same event in all cases.

Update the drawing sample to refresh its system colours scheme whenever
the theme is changed.

vadz commented:
OK, I've implemented this now and also added `wxSysColourChangedEvent` generation on the theme change (still no automatic update of the appearance when this happens though).

Please let me know if anybody has any objections, otherwise I'll finally merge this branch soon.

k-serg commented:
I'd like to see something that return current mode:

```
PreferredAppMode wxMSWDarkMode::GetCurrentMode() const
{
    return gs_appMode;
}
```

Yes, we have `IsActive()` but doesn't know how it was activated.


And a little change wxApp enum:

```
    // MSW-specific function to enable experimental dark mode support.
    enum
    {
        DarkMode_Auto   = 0,  // Use dark mode if the system is using it.
        DarkMode_Always = 1   // Force using dark mode.
    };
```

```
    // MSW-specific function to enable experimental dark mode support.
    enum
    {
        DarkMode_Off   = 0    // Never use dark mode
        DarkMode_Auto   = 1,  // Use dark mode if the system is using it.
        DarkMode_Always = 2   // Force using dark mode.
    };
```

Ofcourse it will be done on my own inside app, as I did already:

```
    enum class DarkMode
    {
        DARK_OFF,
        DARK_AUTO,
        DARK_ALWAYS
    };
```

vadz commented:
Sorry, but why should it matter how it was activated? The only thing which matters is whether the app is using dark mode or not and `wxSystemAppearance::IsDark()` can be used for this and is portable, so I prefer telling people to use it instead of adding another MSW-specific function.

I guess we could add a way to turn off dark mode, but this will only make sense once/if it's ever enabled automatically. As long as it's not the case, I'd rather avoid adding another function when you can just avoid calling `MSWEnableDarkMode()` in the first place. And if we add it, I'd rather call it `MSWDisableDarkMode()` instead of adding another enum element because `MSWEnableDarkMode(DarkMode_Off)` doesn't read well to me.

k-serg commented:
>Sorry, but why should it matter how it was activated?
	
The first thing that comes to mind:

```
	void OnSysColourChangedEvent(wxSysColourChangedEvent& event)
	{
		switch (wxMSWDarkMode::GetCurrentMode() )
		{
			//Should we ignore this, because we force it?
			case PreferredAppMode::AppMode_ForceDark:
			...
			
		}
	}
```

wxSystemAppearance::IsDark() knows nothing about it, and may change independently at runtime. 

>instead of adding another enum element because MSWEnableDarkMode(DarkMode_Off) doesn't read well to me.

I'd prefer consider those enum as actual state. But I admit if someone thinks otherwise.

vadz commented:
I don't know if using `IsDark()` here works currently, i.e. if `ShouldAppsUseDarkMode()` already returns true or false by the time this message is generated (have you tested it already by chance?), but it really should, so that the same code would work under all platforms.

As for distinguishing between forcing dark mode or following the system theme, this is your application option, so you already know it and I don't think wx needs to keep track of it.

k-serg commented:
>I don't know if using IsDark() here works currently, i.e. if ShouldAppsUseDarkMode() already returns true or false by the time this message is generated (have you tested it already by chance?), but it really should, so that the same code would work under all platforms.

At first change apps dark mode:
(event processes 6 times, IDK why)

**Select Dark->Light:**

```
(27.12.2022 22:47:54) [7680] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:47:54) [7680] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
(27.12.2022 22:47:55) [7680] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:47:55) [7680] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
(27.12.2022 22:47:56) [7680] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:47:56) [7680] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
(27.12.2022 22:47:58) [7680] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:47:58) [7680] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
(27.12.2022 22:47:58) [7680] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:47:58) [7680] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
(27.12.2022 22:48:00) [7680] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:48:00) [7680] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
```


**Select Light->Dark:**

```
(27.12.2022 22:44:46) [14228] 'OnSysColourChanged' wxSystemSettings::GetColour: rgb(32, 32, 32)
(27.12.2022 22:44:46) [14228] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (1) wxSystemAppearance::IsDark(): (1)
(27.12.2022 22:44:51) [14228] 'OnSysColourChanged' wxSystemSettings::GetColour: rgb(32, 32, 32)
(27.12.2022 22:44:51) [14228] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (1) wxSystemAppearance::IsDark(): (1)
(27.12.2022 22:44:52) [14228] 'OnSysColourChanged' wxSystemSettings::GetColour: rgb(32, 32, 32)
(27.12.2022 22:44:52) [14228] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (1) wxSystemAppearance::IsDark(): (1)
(27.12.2022 22:44:54) [14228] 'OnSysColourChanged' wxSystemSettings::GetColour: rgb(32, 32, 32)
(27.12.2022 22:44:54) [14228] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (1) wxSystemAppearance::IsDark(): (1)
(27.12.2022 22:44:55) [14228] 'OnSysColourChanged' wxSystemSettings::GetColour: rgb(32, 32, 32)
(27.12.2022 22:44:55) [14228] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (1) wxSystemAppearance::IsDark(): (1)
(27.12.2022 22:44:56) [14228] 'OnSysColourChanged' wxSystemSettings::GetColour: rgb(32, 32, 32)
(27.12.2022 22:44:56) [14228] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (1) wxSystemAppearance::IsDark(): (1)
```

At second and further change apps dark mode:
**Select Dark->Light:**

```
(27.12.2022 22:45:59) [14228] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:45:59) [14228] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
(27.12.2022 22:46:01) [14228] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:46:01) [14228] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
(27.12.2022 22:46:04) [14228] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:46:04) [14228] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
(27.12.2022 22:46:05) [14228] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:46:05) [14228] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
(27.12.2022 22:46:06) [14228] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:46:06) [14228] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
(27.12.2022 22:46:08) [14228] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:46:08) [14228] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
```

**Select Light->Dark (pay attention that first "0" and white):**

```
(27.12.2022 22:48:37) [7680] 'OnSysColourChanged' wxSystemSettings::GetColour: white
(27.12.2022 22:48:37) [7680] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (0) wxSystemAppearance::IsDark(): (0)
(27.12.2022 22:48:38) [7680] 'OnSysColourChanged' wxSystemSettings::GetColour: rgb(32, 32, 32)
(27.12.2022 22:48:38) [7680] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (1) wxSystemAppearance::IsDark(): (1)
(27.12.2022 22:48:41) [7680] 'OnSysColourChanged' wxSystemSettings::GetColour: rgb(32, 32, 32)
(27.12.2022 22:48:41) [7680] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (1) wxSystemAppearance::IsDark(): (1)
(27.12.2022 22:48:42) [7680] 'OnSysColourChanged' wxSystemSettings::GetColour: rgb(32, 32, 32)
(27.12.2022 22:48:42) [7680] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (1) wxSystemAppearance::IsDark(): (1)
(27.12.2022 22:48:43) [7680] 'OnSysColourChanged' wxSystemSettings::GetColour: rgb(32, 32, 32)
(27.12.2022 22:48:43) [7680] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (1) wxSystemAppearance::IsDark(): (1)
(27.12.2022 22:48:44) [7680] 'OnSysColourChanged' wxSystemSettings::GetColour: rgb(32, 32, 32)
(27.12.2022 22:48:44) [7680] 'OnSysColourChanged' wxMSWDarkMode::IsActive(): (1) wxSystemAppearance::IsDark(): (1)
```
Code that you may copy in any sample:
`Connect(wxID_ANY, wxEVT_SYS_COLOUR_CHANGED, (wxObjectEventFunction)&SaveGameProFrame::OnSysColourChanged);`

```
void MyCoolAppFrame::OnSysColourChanged(wxSysColourChangedEvent& event)
{
    bool isActive = wxMSWDarkMode::IsActive();
    bool isDark = wxSystemSettings::GetAppearance().wxSystemAppearance::IsDark();
    wxString col = wxSystemSettings::GetColour(wxSystemColour::wxSYS_COLOUR_WINDOW).GetAsString();

    wxLogDebug("wxSystemSettings::GetColour: %s", col);
    wxLogDebug("wxMSWDarkMode::IsActive(): (%i) wxSystemAppearance::IsDark(): (%i)", isActive , isDark);

    //event.Skip(true);
}
```
results may vary, depend on `event.Skip()`

>As for distinguishing between forcing dark mode or following the system theme, this is your application option, so you already know it and I don't think wx needs to keep track of it.

Well, I have nothing more to add, do what you think is right.

vadz commented:
I've merged it as is for now, we can always add more APIs later (while removing them is much more problematic). Thanks again for your testing and please open new issues for the remaining problems (other than those related to AUI and STC that I'll try to address soon).

eranif commented:
Thanks for the effort invested in this - much appreciated!
I will start using it in `CodeLite` and will open issues if needed

PBfordev commented:
@k-serg  wrote
> At first change apps dark mode: (event processes 6 times, IDK why)

It is odd that so many events are received but it does happen with the current master. In my [wxSystemInformationFrame](https://github.com/PBfordev/wxsysinfoframe) I log some Windows messages and wxWidgets events. When switching the system dark mode for applications from light to dark (after starting the application), I see the message and event 14 times:
```
13:24:05: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:05: wxSysColourChangedEvent received.
13:24:05: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:05: wxSysColourChangedEvent received.
13:24:05: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:05: wxSysColourChangedEvent received.
13:24:05: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:05: wxSysColourChangedEvent received.
13:24:05: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:05: wxSysColourChangedEvent received.
13:24:06: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:06: wxSysColourChangedEvent received.
13:24:06: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:06: wxSysColourChangedEvent received.
13:24:06: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:06: wxSysColourChangedEvent received.
13:24:06: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:06: wxSysColourChangedEvent received.
13:24:06: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:06: wxSysColourChangedEvent received.
13:24:06: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:06: wxSysColourChangedEvent received.
13:24:06: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:06: wxSysColourChangedEvent received.
13:24:06: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:06: wxSysColourChangedEvent received.
13:24:06: WM_SETTINGCHANGE received: wParam = 0, lParam ="ImmersiveColorSet"
13:24:06: wxSysColourChangedEvent received.
```
Not sure if this is expected, or something I do wrong in my `wxSystemInformationFrame` code, or something not correct in wxWidgets.

vadz commented:
FWIW when I toggle between "Light" and "Dark" in the "Colours" section of the "Settings", I can see 5 `WM_SETTINGSCHANGE`s being sent to a trivial Win32-only window in my example using Spy++. 5 is not 14 (as above), but the system clearly does send several messages for a single change and I have no idea why.

We could coalesce multiple consecutive messages into a single one, I guess, but it's still weird that it does it.

k-serg commented:
It seems to me that we should not coalesced messages, we need to store the previous state to the event and then the new one, and compare it in the handler.

k-serg commented:
More interested to me: why on second Light->Dark selection, as I already wrote, I have different values with one message batch.

vadz commented:
Apparently Windows sends the first message before actually changing the colours. I think it's because there are actually several changes as it's an animation and not just an immediate change from white to black, but we don't handle the intermediate messages and just get the "compatible" (?) ones.

PBfordev commented:
> FWIW when I toggle between "Light" and "Dark" in the "Colours" section of the "Settings", I can see 5 `WM_SETTINGSCHANGE`s being sent to a trivial Win32-only window in my example using Spy++. 

I see 14 messages both in Spy++ and the application, using the simplest possible code (Windows 10 22H2):
```cpp
#include <wx/wx.h>

class MyFrame : public wxFrame
{
public:
    MyFrame() : wxFrame(nullptr, wxID_ANY, "Test") {}
private:
    WXLRESULT MSWWindowProc(WXUINT nMsg, WXWPARAM wParam, WXLPARAM lParam) override
    {
        if ( nMsg == WM_SETTINGCHANGE )
            wxLogDebug("WM_SETTINGCHANGE received: wParam = %u, lParam ='%s'", (unsigned)wParam, lParam ? (LPCTSTR)lParam : wxS(""));
        return wxFrame::MSWWindowProc(nMsg, wParam, lParam);
    }
};

class MyApp : public wxApp
{
    bool OnInit() override
    {
        (new MyFrame())->Show();
        return true;
    }
}; wxIMPLEMENT_APP(MyApp);
```

gerard-durand commented:
It could be nice if it was possible to modify some colours in MSW darkmode, for example with something like wxMSWDarkMode::SetColour(wxSYS_COLOUR_MENUTEXT, wxColour(0xe00000)); to obtain, here, a red menu text, as it appears in some astronomy software when you use their ‘night’ mode. The standard mode exists in those softwares and the global look is quite near the darkmode as it is now, with menus in white (or light gray).
This should be probably called at the beginning of a program, before the build of menus.

vadz commented:
@gerard-durand FYI, this is possible now, see #23275, notably `wxDarkModeSettings::GetMenuColour()`. Any feedback on that PR is welcome, please comment there if you'd like to help improving it before it's merged. TIA!

vadz commented:
The example from the documentation should work, i.e. just
```cpp
    class MySettings : public wxDarkModeSettings
    {
    public:
        virtual wxColour GetMenuColour(wxMenuColour which)
        {
            if ( which == wxMenuColour::StandardFg )
                return *wxRED;

            return wxDarkModeSettings::GetMenuColour(which);
        }
    };

    wxTheApp->MSWEnableDarkMode(wxApp::DarkMode_Always, new MySettings());
```
will use red for the top level menus in dark mode.

Customizing normal menu colours is shown in the ownerdrw sample.

gerard-durand commented:
> Customizing normal menu colours is shown in the ownerdrw sample.
>
Ok for a single item, but how to set the color globally for all sub-menus items.

vadz commented:
There is no support for changing this colour globally, sorry.

The goal of these changes was to just allow customizing the hardcoded colours used in the dark mode because they can be wrong.

gerard-durand commented:
I have tested your suggested code in the sample ownerdrw.
I have just added : 
if (wxSystemSettings::GetAppearance().IsDark()) wxTheApp->MSWEnableDarkMode(wxApp::DarkMode_Always, new MySettings());
before the call to InitMenu();
To obtain the darkmode, I launch ownerdraw.exe in a cmd file containing :

set WX_MSW_DARK_MODE=2
ownerdrw.exe

Like that, only the last sub-menu (Native) has correct colors in darkmode (black background and white text).
To obtain black backgrounds on submenus, I had to comment lines containing :

pItem->SetTextColour`
pItem->SetFont only for sub-menus
and
pAboutItem->SetDisabledBitmap

>There is no support for changing this colour globally, sorry.
>
Will it be possible in the future ? Because, at least internally, this is done in darkmode with white sub-menus on a black background.

vadz commented:
I'm surprised that "File" and "Drawn" menus don't use the correct colours. Could you please open a new issue with the exact diff that you use so that I could test it?

> Will it be possible in the future ? Because, at least internally, this is done in darkmode with white sub-menus on a black background.

No, normal menus are drawn in dark mode by Windows itself. It doesn't draw the menu bar correctly in the dark mode, which is why we have to use our own hack for it.

gerard-durand commented:
Here is a diff file which is OK for me (with commented lines)
[ownerdrw.zip](https://github.com/wxWidgets/wxWidgets/files/10794779/ownerdrw.zip)

gerard-durand commented:
Just to add something. If I restore only one attribute commented for a sub-menu, it's all the sub-menu which is displayed with standard colors (not darkmode).

gerard-durand commented:
New remark : If as in the doc of wxDarkModeSettings Class Reference, I use the switch as it is, I obtain many warnings of my compiler (msys2 12.2) like this one:
warning: enumeration value 'wxSYS_COLOUR_SCROLLBAR' not handled in switch [-Wswitch]
If I add a default case like:

    switch ( index )
        {
            ...   
            // Default colour used here is 0xe0e0e0.
            return *wxRED;

        default:
            return wxDarkModeSettings::GetColour(index);

        }

It works and I have no compiler warning.

vadz commented:
Sorry for the delay with replying, but, in any case, I'm afraid I can't say anything really useful: it's clear that owner-drawn menu items can't be used in dark mode, so I can only document this limitation (which I will do, and will also fix the example to avoid triggering `-Wswitch`, thanks).

gerard-durand commented:
Thanks for your (whole) work.
I have also the feeling that this darkmode problem for sub-menus is not simple. I use commonly Notepad++. They have a darkmode which is customisable for the main menu and/or main toolbar, but not for sub-menus which are still white on a black background. And, they don't use wxWidgets, but directly the Windows API. So, solving that needs probably to dig deeply inside Windows. :-(
Nevertheless, it's possible to set independently colors for each submenu item, but separators are not affected, there is still a white large clear border around the sub-menus. Not a nice look. May be we'll have to wait for a next Windows correction...

gerard-durand commented:
I continue to test darkmode on Windows.
wxAboutBox, in simple mode as seen in samples\dialogs\dialogs.cpp, is shown with standard Windows colors, even when started with a preceeding set WX_MSW_DARK_MODE=2. Normal: because in this case you use the standard Windows box. All other "about" tests are shown in dark.
In that example, it is possible to force the usage of wxGenericAboutBox in ShowSimpleAboutDialog, line 3458, and it works.
When I try this in my own work, it does not compile because with only #include "wx/aboutdlg.h", wxGenericAboutBox is not found/declared. I have to add explicitely #include "wx/generic/aboutdlgg.h".
Note that in interface/wx/aboutdlg.h, wxGenericAboutBox is declared, but if I understand well, this interface folder is only used for documentation.

It can work in my own work with a simple wxAboutBox if I add a aboutInfo.SetWebSite. In that case, wxAboutBox switches automatically and internally to wxGenericAboutBox.
Nevertheless, in the documentation of wxGenericAboutBox, in one of the windows I have found, it is just stated that we have to use #include "wx/aboutdlg.h":
> wxGenericAboutBox()
> void wxGenericAboutBox  ( const wxAboutDialogInfo &  info,  
>   wxWindow *  parent = nullptr  
>  )   
> This function does the same thing as wxAboutBox() except that it always uses the generic wxWidgets version of the dialog instead of the native one. 
> This is mainly useful if you need to customize the dialog by e.g. adding custom controls to it (customizing the native dialog is not currently supported).
> See the Dialogs Sample for an example of about dialog customization.
> See also wxAboutDialogInfo
> Include file: 
> #include <wx/aboutdlg.h> 

In an other one, it is just stated that we have to use #include "wx/generic/aboutdlgg.h".
But, apparently, both includes are necessary to be able to use directly wxGenericAboutBox, and in my case, obtain a dark about window (even without adding an url).

gerard-durand commented:
**About wxProgressDialog**:
On Windows, wxProgressDialog does not support darkmode, but wxGenericProgressDialog supports it. You can see this in the samples/dialogs if you set WX_MSW_DARK_MODE=2 before launching dialogs.exe.

PBfordev commented:
Many (but not all) limitations of MSW dark mode (stemming from the lack of dark mode support in the native controls) are documented, including `wxAboutBoxBox()` and `wxProgressDialog`, see https://docs.wxwidgets.org/latest/classwx_app.html#af8c93d7e3345e62a58325f3ab1d158d6

Some others are known and reported, so far without a fix...
